Share experiment with copy/email buttons

ulziibay opened the pull request:
<img width="297" alt="screen shot 2018-08-29 at 5 00 18 pm" src="https://user-images.githubusercontent.com/1626382/44874657-be9b2580-ac50-11e8-9e78-d14bbb6ada34.png">

lediur commented on src/content/Discussion.tsx:
```diff
@@ -1,22 +1,20 @@
+import "./Discussion.css";
```
Can you set up your editor so it will auto sort and organize imports and configure it to run on save? If I touch this file in the future it will automatically sort these imports and cause a lot of churn.

ulziibay replied:
Yeah for atom, i was using https://atom.io/packages/atom-import-sort but its not TSLint compliant. I just run `tsling --fix`. AFAIK, this version is compatible with tslint order rule!

lediur replied:
I don't think I'm using the tslint fixer - I'm using the VSCode auto-ordering and sort thing, which should be provided by the typescript language service https://blogs.msdn.microsoft.com/typescript/2018/03/27/announcing-typescript-2-8/#organize-imports

lediur commented on src/content/Discussion.tsx:
```diff
@@ -110,9 +108,7 @@ class CommentsWell extends React.PureComponent<CommentsWellProps> {
-          {comments.map((comment, i) => (
-            <CommentItem key={i} {...comment} />
-          ))}
+          {comments.map((comment, i) => <CommentItem key={i} {...comment} />)}
```
I think this conflicts with Prettier configuration - can you take a look

lediur commented on src/content/RepoTwist.tsx:
```diff
@@ -13,6 +13,7 @@ import {
   permissionsUrl
 } from "@r2c/extension/api/permissions";
 import { scoreRepoUrl, ScoreResponse } from "@r2c/extension/api/score";
+import { CopyButton } from "@r2c/extension/shared/CopyButton";
```
For components that have their own files, can you make those the default export?

lediur replied:
e.g. `import CopyButton from "@r2c/extension/shared/CopyButton"`, and `export default class CopyButton` if it's a class / `const CopyButton: React.SFC…` and `export default CopyButton` if it's an SFC

ulziibay replied:
WIll do!

lediur commented on src/content/RepoTwist.tsx:
```diff
@@ -153,13 +141,15 @@ class RepoPackageSection extends React.Component<{}, RepoPackageSectionState> {
+                                user={endorser}
+                                className="package-endorser"
+                              />
+                            ))}
```
I think this conflicts with Prettier

ulziibay replied:
I am crossing checking atom-prettier with VS code prettier

lediur commented on src/content/Share.tsx:
```diff
@@ -0,0 +1,110 @@
+  hovered: boolean;
+}
+
+const EmailIcon: React.SFC<EmailIconProps> = ({ hovered }) => (
```
What do you use `hovered` for? I think these props are unused

lediur replied:
If you want to do hover state let's use CSS there

ulziibay replied:
Agreed!

lediur commented on src/content/Share.tsx:
```diff
@@ -0,0 +1,110 @@
+          <h1 className="twist-title">Share R2C!</h1>
+        </header>
+        <div className="twist-body">{this.props.shortDesc}</div>
+        <div>
```
nit: let's avoid naked elements with no class names - makes it difficult to understand layout and semantic intent

ulziibay replied:
Will do!

lediur commented on src/content/Share.tsx:
```diff
@@ -0,0 +1,110 @@
+  private onEmailLinkClick: React.MouseEventHandler<HTMLElement> = e => {
+    const { onEmailClick } = this.props;
+    if (onEmailClick) {
+      onEmailClick(e);
```
Since this click handler is already bound to a link, do you _also_ want it to copy the link to the clipboard?

ulziibay replied:
Sure!

lediur commented on src/content/index.tsx:
```diff
@@ -208,4 +265,10 @@ export default class ContentHost extends React.Component<{}, ContentHostState> {
+  private onShareActionClick = (
+    buttonTitle: ShareActionType
+  ): React.MouseEventHandler<HTMLElement> => e => {
+    console.log("Logging share link aciton!");
```
spelling

lediur commented on tslint.json:
```diff
@@ -29,6 +30,7 @@
     "no-void-expression": [true, "ignore-arrow-function-shorthand"],
     "no-submodule-imports": false,
     "max-func-body-length": false,
+    "max-classes-per-file": [true, 5],
```
I'm not sure I agree with this - I usually decompose things a lot into separate SFCs and this would cause overhead there

ulziibay replied:
Was triggering on index.tsx. Will undo this and adhere to not more than 3. It could be strict no more than 1 as well!

lediur replied:
yeah I think a "gut check" here is more appropriate than a tslint check

lediur replied:
My bad - I thought the default was set to `false` originally, not `[true, 3]`, so I thought you were adding this restriction to the linter. This change seems fine then.

lediur commented on src/content/Share.css:
```diff
@@ -0,0 +1,19 @@
+.r2c-twist .twist .share-twist {
```
I don't think this actually applies since the component itself has both `twist` and `share-twist` class names. I think you want `.twist.share-twist`

ulziibay replied:
It does not. Will remove!

lediur commented on src/content/Share.tsx:
```diff
@@ -0,0 +1,110 @@
+  ShareSectionState
+> {
+  public state: ShareSectionState = {
+    clickedEmail: false,
```
This state seems unused

ulziibay replied:
Was going to use this to trigger logging! But the way it set up, it will log everything, which is what I want here. Will remove state!

lediur commented on src/shared/CopyButton.tsx:
```diff
@@ -0,0 +1,24 @@
+  </svg>
+);
+
+export const CopyButton: React.SFC<CopyButtonProps> = ({ onClick }) => (
```
`const CopyButton`, then later `export default CopyButton`

(you can't do `export default const CopyButton` for "reasons")

ulziibay replied:
Alright!

ulziibay commented:
@lediur: Addressed all comments except import ordering. AFAIK, without changing to VScode, atom does not have a good import order module other than the one I linked above. And there is no CLI from TS checker for ordering the imports.
Put the share button in the middle. 
<img width="302" alt="screen shot 2018-09-04 at 10 35 34 am" src="https://user-images.githubusercontent.com/1626382/45047582-49539a00-b02e-11e8-9921-8f74d87095fa.png">

lediur commented:
only other comment: can we change the "Share our extension with your friends..." text to "Share the extension with your friends using the options below"
